File: Linked List/doubly_circular/deletion/list.cpp
// Doubly Circular Linked List - Deletion (C++)
// Delete head and first occurrence by value.

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* prev;
    Node* next;
    Node(int value) : data(value), prev(nullptr), next(nullptr) {}
};

Node* deleteHead(Node* head) {
    if (head == nullptr) return nullptr;
    if (head->next == head) { delete head; return nullptr; }
    Node* tail = head->prev;
    Node* newHead = head->next;
    tail->next = newHead; newHead->prev = tail;
    delete head;
    return newHead;
}

Node* deleteByValue(Node* head, int target) {
    if (head == nullptr) return nullptr;
    Node* cur = head;
    do {
        if (cur->data == target) {
            if (cur == head) return deleteHead(head);
            cur->prev->next = cur->next;
            cur->next->prev = cur->prev;
            delete cur;
            return head;
        }
        cur = cur->next;
    } while (cur != head);
    return head;
}

void printForward(Node* head) {
    if (head == nullptr) { cout << "\n"; return; }
    Node* cur = head; do { cout << cur->data << " "; cur = cur->next; } while (cur != head); cout << "\n";
}

int main() {
    // Build circular: 10 <-> 20 <-> 30
    Node* n1 = new Node(10); Node* n2 = new Node(20); Node* n3 = new Node(30);
    n1->next = n2; n2->prev = n1; n2->next = n3; n3->prev = n2; n3->next = n1; n1->prev = n3;
    Node* head = n1;
    cout << "Initial: "; printForward(head);
    head = deleteHead(head); cout << "After deleteHead: "; printForward(head);
    head = deleteByValue(head, 30); cout << "After deleteByValue(30): "; printForward(head);
    return 0;
}

File: Linked List/doubly/deletion/list.py
# Doubly Linked List - Deletion (Python)
# Delete head and first occurrence by value.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.prev: "Node | None" = None
        self.next: "Node | None" = None


def delete_head(head: "Node | None") -> "Node | None":
    if head is None:
        return None
    new_head = head.next
    if new_head is not None:
        new_head.prev = None
    return new_head


def delete_by_value(head: "Node | None", target: int) -> "Node | None":
    cur = head
    while cur is not None and cur.data != target:
        cur = cur.next
    if cur is None:
        return head
    if cur.prev is not None:
        cur.prev.next = cur.next
    else:
        head = cur.next
    if cur.next is not None:
        cur.next.prev = cur.prev
    return head


def print_forward(head: "Node | None") -> None:
    cur = head
    while cur is not None:
        print(cur.data, end=" ")
        cur = cur.next
    print()


if __name__ == "__main__":
    # Build: 10 <-> 20 <-> 30 <-> 40
    n1 = Node(10); n2 = Node(20); n3 = Node(30); n4 = Node(40)
    n1.next = n2; n2.prev = n1; n2.next = n3; n3.prev = n2; n3.next = n4; n4.prev = n3
    head = n1

    print("Initial:", end=" "); print_forward(head)
    head = delete_head(head); print("After delete_head:", end=" "); print_forward(head)
    head = delete_by_value(head, 30); print("After delete_by_value(30):", end=" "); print_forward(head)


